File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/BST/lbh_n-ary_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.children = []

# Function to add a child to a node
def add_child(parent, data):
    parent.children.append(Node(data))

# Function to delete a node (assumes parent is known)
def delete_node(parent, data):
    for i, child in enumerate(parent.children):
        if child.data == data:
            del parent.children[i]
            return
    for child in parent.children:
        delete_node(child, data)

# Function for breadth-first traversal
def bft(root):
    if not root:
        return
    queue = [root]
    while queue:
        node = queue.pop(0)
        print(node.data, end=' ')
        queue.extend(node.children)

# Function for preorder traversal
def preorder(root):
    if not root:
        return
    print(root.data, end=' ')
    for child in root.children:
        preorder(child)

# Function for postorder traversal
def postorder(root):
    if not root:
        return
    for child in root.children:
        postorder(child)
    print(root.data, end=' ')

# Function to search a node
def search(root, data):
    if not root:
        return None
    if root.data == data:
        return root
    for child in root.children:
        result = search(child, data)
        if result:
            return result
    return None

# Example usage
root = Node(1)
add_child(root, 2)
add_child(root, 3)
add_child(root, 4)

child = search(root, 2)
if child:
    add_child(child, 5)
    add_child(child, 6)

child = search(root, 3)
if child:
    add_child(child, 7)

print("Breadth-First Traversal:", end=' ')
bft(root)
print()

print("Preorder Traversal:", end=' ')
preorder(root)
print()

print("Postorder Traversal:", end=' ')
postorder(root)
print()

print("\nDeleting node 2 and its children")
delete_node(root, 2)

print("Breadth-First Traversal:", end=' ')
bft(root)
print()

File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/BST/lbh_n-ary_tree.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Node {
public:
    int data;
    vector<Node*> children;
    Node(int val) : data(val) {}
};

// Function to add a child to a node
void addChild(Node* parent, int data) {
    parent->children.push_back(new Node(data));
}

// Function to delete a node (assumes parent is known)
void deleteNode(Node* parent, int data) {
    for (auto it = parent->children.begin(); it != parent->children.end(); ++it) {
        if ((*it)->data == data) {
            delete *it;
            parent->children.erase(it);
            return;
        }
    }
    for (auto child : parent->children) {
        deleteNode(child, data);
    }
}

// Function for breadth-first traversal
void bft(Node* root) {
    if (!root) return;
    queue<Node*> q;
    q.push(root);
    while (!q.empty()) {
        Node* node = q.front();
        q.pop();
        cout << node->data << " ";
        for (auto child : node->children) {
            q.push(child);
        }
    }
}

// Function for preorder traversal
void preorder(Node* root) {
    if (!root) return;
    cout << root->data << " ";
    for (auto child : root->children) {
        preorder(child);
    }
}

// Function for postorder traversal
void postorder(Node* root) {
    if (!root) return;
    for (auto child : root->children) {
        postorder(child);
    }
    cout << root->data << " ";
}

// Function to search a node
Node* search(Node* root, int data) {
    if (!root) return nullptr;
    if (root->data == data) return root;
    for (auto child : root->children) {
        Node* result = search(child, data);
        if (result) return result;
    }
    return nullptr;
}

int main() {
    Node* root = new Node(1);
    addChild(root, 2);
    addChild(root, 3);
    addChild(root, 4);

    Node* child = search(root, 2);
    if (child) {
        addChild(child, 5);
        addChild(child, 6);
    }

    child = search(root, 3);
    if (child) {
        addChild(child, 7);
    }

    cout << "Breadth-First Traversal: ";
    bft(root);
    cout << endl;

    cout << "Preorder Traversal: ";
    preorder(root);
    cout << endl;

    cout << "Postorder Traversal: ";
    postorder(root);
    cout << endl;

    cout << "\nDeleting node 2 and its children\n";
    deleteNode(root, 2);

    cout << "Breadth-First Traversal: ";
    bft(root);
    cout << endl;

    return 0;
}